ksqldb-engine/src/main/java/io/confluent/ksql/format/DefaultFormatInjector.java

```diff
@@ -0,0 +1,190 @@
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.KsqlStatementException;
+import java.util.Optional;
+
+/**
+ * An injector which injects the key and value formats into the supplied {@code statement}.
+ *
+ * <p>The key format is only injected if:
+ * <ul>
+ * <li>The statement is a CT/CS.</li>
+ * <li>The statement does not specify the FORMAT property in its WITH clause.</li>
+ * <li>The statement does not specify the KEY_FORMAT property in its WITH clause.</li>
+ * </ul>
+ *
+ * <p>Similarly, the value format is only injected if the above conditions are met,
+ * where the KEY_FORMAT property is replaced with the VALUE_FORMAT property accordingly.
+ *
+ * <p>If any of the above are not true then the {@code statement} is returned unchanged.
+ */
+public class DefaultFormatInjector implements Injector {
+
+  public DefaultFormatInjector() {
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T extends Statement> ConfiguredStatement<T> inject(
+      final ConfiguredStatement<T> statement
+  ) {
+    if (featureFlagNotEnabled(statement)) {
+      validateLegacyFormatProperties(statement);
+    }
+
+    if (statement.getStatement() instanceof CreateSource) {
+      return handleCreateSource((ConfiguredStatement<CreateSource>) statement);
+    }
+
+    return statement;
+  }
+
+  @SuppressWarnings("unchecked")
+  private <T extends Statement> ConfiguredStatement<T> handleCreateSource(
+      final ConfiguredStatement<CreateSource> statement
+  ) {
+    try {
+      // Safe to cast as we know `T` is `CreateSource`
+      return (ConfiguredStatement<T>)
+          injectForCreateStatement(statement).orElse(statement);
+    } catch (final KsqlStatementException e) {
+      throw e;
+    } catch (final KsqlException e) {
+      throw new KsqlStatementException(
+          ErrorMessageUtil.buildErrorMessage(e),
+          statement.getStatementText(),
+          e.getCause());
+    }
+  }
+
+  private Optional<ConfiguredStatement<CreateSource>> injectForCreateStatement(
+      final ConfiguredStatement<CreateSource> original
+  ) {
+
+    final CreateSource statement = original.getStatement();
+    final CreateSourceProperties properties = statement.getProperties();
+
+    final Optional<FormatInfo> keyFormat = properties.getKeyFormat();
+    final Optional<FormatInfo> valueFormat = properties.getValueFormat();
+
+    if (keyFormat.isPresent() && valueFormat.isPresent()) {
+      return Optional.empty();
+    }
+
+    final KsqlConfig config = getConfig(original);
+    validateConfig(config, keyFormat, valueFormat);
+
+    final CreateSourceProperties injectedProps = properties.withFormats(
+        keyFormat.map(FormatInfo::getFormat)
+            .orElse(config.getString(KsqlConfig.KSQL_DEFAULT_KEY_FORMAT_CONFIG)),
+        valueFormat.map(FormatInfo::getFormat)
+            .orElse(config.getString(KsqlConfig.KSQL_DEFAULT_VALUE_FORMAT_CONFIG))
+    );
```

Comment:
nit:  this code is correct and works.  However, just a suggestion in terms of the pattern...

This code separates the validation from the access.  You first validate the state is what you need, then you access the state.  This works well until someone comes along and changes one side without realizing they should update the other.

I'd suggest a better pattern is to access the state and validate what you extract. This makes it much harder for someone to come along and miss something, e.g. 

```java
private Optional<ConfiguredStatement<CreateSource>> injectForCreateStatement(
      final ConfiguredStatement<CreateSource> original
  ) {
   ....
 
   final KsqlConfig config = getConfig(original);

    final CreateSourceProperties injectedProps = properties.withFormats(
        keyFormat.map(FormatInfo::getFormat).orElseGet(() -> getDefaultKeyFormat(config)),
        valueFormat.map(FormatInfo::getFormat).orElseGet(() -> getDefaultValueFormat(config))
    );

    ...
 }

 private String getDefaultKeyFormat(final KsqlConfig config) {
     final String format = config.getString(KsqlConfig.KSQL_DEFAULT_KEY_FORMAT_CONFIG);
    if (format == null) {
        throw new KsqlException("Statement is missing the '"
          + CommonCreateConfigs.KEY_FORMAT_PROPERTY + "' property from the WITH clause. "
          + "Either provide one or set a default via the '"
          + KsqlConfig.KSQL_DEFAULT_KEY_FORMAT_CONFIG + "' config.");
    }

    return format;
 }

 // Same pattern for getDefaultValueFormat
```

Here, both the accessing and validating of the default key format is done in one place: the `getDefaultKeyFormat` method.

Reply:
Great tip! Made the change, will also keep this in mind for the future.